pkg/sql/opt/memo/statistics_builder.go

```diff
@@ -326,25 +332,59 @@ func (sb *statisticsBuilder) colStatLeaf(
 	}
 
 	// If some of the columns are a lax key, the distinct count equals the row
-	// count. Note that this doesn't take into account the possibility of
-	// duplicates where all columns are NULL.
+	// count. The null count is 0 if these columns are not nullable, otherwise
+	// copy the null count from the nullable columns in colSet.
 	if fd.ColsAreLaxKey(colSet) {
-		colStat.DistinctCount = s.RowCount
+		if colSet.SubsetOf(notNullCols) {
+			colStat.NullCount = 0
+		} else {
+			nullableCols := colSet.Difference(notNullCols)
+			if nullableCols.Equals(colSet) {
+				// No column statistics on this colSet - use the unknown
+				// null count ratio.
+				colStat.NullCount = s.RowCount * unknownNullCountRatio
+			} else {
+				colStatLeaf := sb.colStatLeaf(nullableCols, s, fd, notNullCols)
+				colStat.NullCount = colStatLeaf.NullCount
+			}
+		}
+		colStat.DistinctCount = s.RowCount - colStat.NullCount
 		return colStat
 	}
 
 	if colSet.Len() == 1 {
 		col, _ := colSet.Next(0)
 		colStat.DistinctCount = unknownDistinctCountRatio * s.RowCount
+		colStat.NullCount = unknownNullCountRatio * s.RowCount
 		if sb.md.ColumnType(opt.ColumnID(col)) == types.Bool {
 			colStat.DistinctCount = min(colStat.DistinctCount, 2)
 		}
+		if notNullCols.Contains(col) {
+			colStat.NullCount = 0
+		}
 	} else {
 		distinctCount := 1.0
+		nullCount := 0.0
+		// The collision factor tracks expected cases of rows containing
+		// multiple nulls across the reflected columns. If we do a simple
+		// sum of all null counts, the result overestimates the count by
+		// this factor multiplied with RowCount.
+		nullCountCollisionFactor := 1.0
 		colSet.ForEach(func(i int) {
-			distinctCount *= sb.colStatLeaf(util.MakeFastIntSet(i), s, fd).DistinctCount
+			colStatLeaf := sb.colStatLeaf(util.MakeFastIntSet(i), s, fd, notNullCols)
+			distinctCount *= colStatLeaf.DistinctCount
+			nullCount += colStatLeaf.NullCount
+			// This collision factor works okay for 2-column
+			// cases, but starts to really underestimate collisions for
+			// larger sets of columns.
+			//
+			// TODO(itsbilal): Come up with a reasonable average-case
+			// calculation of null collisions across
+			nullCountCollisionFactor *= colStatLeaf.NullCount / s.RowCount
```

Comment:
It's underestimating with larger sets because `nullCountCollisionFactor` gets smaller and smaller as you add more columns. Ideally it should get larger and larger....

My intuition is that it would be easier to just collect the expected number of collisions on each round. So you can just update the `nullCount` calculation above to be:
```
if nullCount < s.RowCount {
  nullCount += colStatLeaf.NullCount * (1 - nullCount / s.RowCount)
}
```
Does that look right?

Reply:
This makes sense. Made the change - thanks!